### PR TITLE
Progressive distillation

### PR DESCRIPTION
**Pull request by bglick13**
Starting to work on implementing [PROGRESSIVE DISTILLATION FOR FAST SAMPLING OF DIFFUSION MODELS](https://arxiv.org/pdf/2202.00512.pdf)

This is very much a draft PR. So far I've included a toy example in a notebook. 

First, it trains an unconditional image diffusion model on a single image. Then, it implements the distillation training procedure - training a student model to produce the same output in N // 2 steps. 

TODOs include:

- [ ] Verify the equation block in the distillation function is correct
- [ ] Test on multiple iterations of distillation (i.e., N = 1000 -> 500 -> 250...)
- [ ] Test on a larger image dataset
- [ ] Clean up changes in DDIM scheduler
- [ ] Convert distillation process to a pipeline(?) or script

### CODE REVIEW

**Comment by natolambert** [comment]
This is a great start. I noticed a couple small bugs and created a [colab](https://colab.research.google.com/drive/1M1vdP6mFBxVTLbIHrNhWtfLTKO4jaegy?usp=sharing) to play with the sampling notebook.

The core code from algorithm 2 should be updated like this (the calculation of `t'` and `t''` was swapped).

```
# get alpha and sigma for base step
alpha_t, sigma_t = teacher_scheduler.get_alpha_sigma(batch, timesteps + 1, accelerator.device)
                
# base output
 z_t = alpha_t * batch + sigma_t * noise
                
# get alpha and sigma for skip steps
alpha_t_prime, sigma_t_prime = student_scheduler.get_alpha_sigma(batch, timesteps // 2, accelerator.device)
alpha_t_prime2, sigma_t_prime2 = teacher_scheduler.get_alpha_sigma(batch, timesteps, accelerator.device)
                
# reconstruction base sample
v = teacher(z_t.float(), timesteps + 1).sample
rec_t = (alpha_t * z_t - sigma_t * v).clip(-1, 1)

# distillation computation
z_t_prime = alpha_t_prime * rec_t + (sigma_t_prime / sigma_t) * (z_t - alpha_t * rec_t)
v_1 = teacher(z_t_prime.float(), timesteps).sample
rec_t_prime = (alpha_t_prime * z_t_prime - sigma_t_prime * v_1).clip(-1, 1)
z_t_prime_2 = alpha_t_prime2 * rec_t_prime + (sigma_t_prime2 / sigma_t_prime) * (z_t_prime - alpha_t_prime * rec_t_prime)
                
# teacher target 
x_hat = z_t_prime_2 - ((sigma_t_prime2 / sigma_t_prime) * z_t) / (alpha_t_prime2 - (sigma_t_prime2 / sigma_t_prime) * alpha_t)
```

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 1]
```diff
@@ -27,6 +27,13 @@
 from .scheduling_utils import SchedulerMixin
 
 
+def E_(input, t, shape, device):
```
Comment (natolambert) [inline src/diffusers/schedulers/scheduling_ddim.py:30]: let's give this a better name, remove 1 letter variables, and maybe a short comment on use.

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 2]
```diff
@@ -138,6 +145,9 @@ def __init__(
 
         self.alphas = 1.0 - self.betas
         self.alphas_cumprod = torch.cumprod(self.alphas, dim=0)
+        self.sqrt_alphas_cumprod = torch.sqrt(self.alphas_cumprod)
+        self.sqrt_one_minus_alphas_cumprod = torch.sqrt(1 - self.alphas_cumprod)
+        self.one = torch.tensor(1.0)
```
Comment (natolambert) [inline src/diffusers/schedulers/scheduling_ddim.py:150]: `self.one` can move into `SchedulerMixin`

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 3]
```diff
@@ -308,3 +318,8 @@ def add_noise(
 
     def __len__(self):
         return self.config.num_train_timesteps
+
+    def get_alpha_sigma(self, x, t, device):
```
Comment (natolambert) [inline src/diffusers/schedulers/scheduling_ddim.py:322]: no 1 letter variables.

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 4]
```diff
@@ -26,6 +26,13 @@
 from .scheduling_utils import SchedulerMixin
 
 
+def E_(input, t, shape, device):
```
Comment (natolambert) [inline src/diffusers/schedulers/scheduling_ddpm.py:29]: same as above

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 5]
```diff
@@ -309,3 +322,8 @@ def add_noise(
 
     def __len__(self):
         return self.config.num_train_timesteps
+
+    def get_alpha_sigma(self, x, t, device):
```
Comment (natolambert) [inline src/diffusers/schedulers/scheduling_ddpm.py:326]: same as above

**Review by natolambert — COMMENTED** [review]
Some little comments. The pipeline looks solid, I think we should test it further.

**Comment by lukovnikov** [comment]
I was thinking of implementing this too :)
I briefly went over the code and I'm wondering if you really want to be training (distilling) inside a Pipeline? From what I understood, it seems pipelines are for sampling?

**Comment by bglick13** [comment]
@lukovnikov sorry for the delayed response. I think that's an important meta decision to make with the Diffusers team! I've seen a similar approach e.g., in the Imagic PR. For now I'm still working on getting the distillation results as good as the paper's!

**Comment by lukovnikov** [comment]
@bglick13 I've also implemented distillation in my fork: https://github.com/lukovnikov/diffusers/blob/mine/examples/unconditional_image_generation/distill_unconditional.py
However, I'm facing some strange issues where the contrast of the generated images is increased after every distillation phase. Did you have similar issues and did you manage to reproduce the paper's results?

**Comment by Thomas-MMJ** [comment]
lukovnikov - at a guess the text encoder tokens are also being trained and are becoming 'longer' which causes the images to be overbaked.

**Comment by adhikjoshi** [comment]
@patrickvonplaten do you think this is something to give it a try again?
